vendor/bat-native-ledger/src/bat/ledger/internal/promotion/promotion_transfer.cc

```diff
@@ -0,0 +1,174 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+
+#include "base/json/json_writer.h"
+#include "base/values.h"
+#include "bat/ledger/internal/contribution/contribution_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/promotion/promotion_transfer.h"
+#include "bat/ledger/internal/promotion/promotion_util.h"
+#include "bat/ledger/internal/request/promotion_requests.h"
+#include "bat/ledger/internal/request/request_util.h"
+#include "bat/ledger/internal/state_keys.h"
+#include "net/http/http_status_code.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+namespace braveledger_promotion {
+
+PromotionTransfer::PromotionTransfer(bat_ledger::LedgerImpl* ledger) :
+ledger_(ledger) {
+  DCHECK(ledger_);
+}
+
+PromotionTransfer::~PromotionTransfer() = default;
+
+void PromotionTransfer::Start(
+    ledger::ExternalWalletPtr wallet,
+    ledger::ResultCallback callback) {
+  // we only need to call old anon api once
+  if (ledger_->GetBooleanState(ledger::kStateAnonTransferChecked)) {
+    GetEligibleTokens(callback);
+    return;
+  }
+
+  if (!wallet) {
+    BLOG(ledger_, ledger::LogLevel::LOG_ERROR) << "Wallet is empty";
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+
+  auto transfer_callback = std::bind(&PromotionTransfer::OnAnonExternalWallet,
+      this,
+      _1,
+      callback);
+
+  ledger_->TransferAnonToExternalWallet(
+      std::move(wallet),
+      transfer_callback,
+      true);
+}
+
+void PromotionTransfer::OnAnonExternalWallet(
+    const ledger::Result result,
+    ledger::ResultCallback callback) {
+  if (result != ledger::Result::LEDGER_OK) {
+    BLOG(ledger_, ledger::LogLevel::LOG_ERROR) << "Initial transfer failed";
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+
+  ledger_->SetBooleanState(ledger::kStateAnonTransferChecked, true);
+  GetEligibleTokens(callback);
+}
+
+void PromotionTransfer::GetEligibleTokens(ledger::ResultCallback callback) {
+  auto tokens_callback = std::bind(&PromotionTransfer::SendTokens,
+      this,
+      _1,
+      callback);
+
+  ledger_->GetUnblindedTokensByPromotionType(
+      GetEligiblePromotions(),
+      tokens_callback);
+}
+
+void PromotionTransfer::SendTokens(
+    ledger::UnblindedTokenList list,
+    ledger::ResultCallback callback) {
+  if (list.empty()) {
+    callback(ledger::Result::LEDGER_OK);
+    return;
+  }
+
+  base::Value credentials(base::Value::Type::LIST);
+  std::vector<std::string> tokens_ids;
+
+  for (auto& item : list) {
+    if (!item) {
+      continue;
+    }
+
+    base::Value token(base::Value::Type::DICTIONARY);
+    bool success = false;
+    if (ledger::is_testing) {
```

Comment:
Since this logic is also in `contribution_util.cc`, I wonder if there might be a way to make this a single call, either by passing in `ledger::is_testing` or something similar.